tickit/adapters/servers/http.py

```diff
@@ -0,0 +1,82 @@
+# import json
+import asyncio
+import logging
+from typing import AsyncIterable, Awaitable, Callable
+
+from aiohttp import web
+
+# from tickit.adapters.interpreters.endpoints.rest_endpoint import RestEndpoint
+from tickit.core.adapter import ConfigurableServer
+from tickit.utils.byte_format import ByteFormat
+
+# from asyncio.streams import StreamReader, StreamWriter
+# from typing import List
+
+
+LOGGER = logging.getLogger(__name__)
+
+routes = web.RouteTableDef()
+
+
+class HttpServer(ConfigurableServer):
+    """A configurable http server with message handling for use in adapters."""
+
+    def __init__(
+        self,
+        host: str = "localhost",
+        port: int = 8080,
+        format: ByteFormat = ByteFormat(b"%b"),
+    ) -> None:
+        """The HttpServer constructor which takes a host, port and format byte string.
+
+        Args:
+            host (str): The host name which the server should be run under.
+            port (int): The port number which the server should listen to.
+            format (ByteFormat): A formatting string for messages sent by the server,
+                allowing for the prepending and appending of data. Defaults to b"%b".
+        """
+        self.host = host
+        self.port = port
+        self.format = format.format
+
+    async def run_forever(
+        self,
+        on_connect: Callable[[], AsyncIterable[bytes]],
+        handler: Callable[[bytes], Awaitable[AsyncIterable[bytes]]],
+    ) -> None:
+        """Runs the HTTP server indefinitely on the configured host and port.
+
+        An asynchronous method used to run the server indefinitely on the configured
+        host and port.
+
+        Args:
+            on_connect (Callable[[], AsyncIterable[bytes]]): An asynchronous iterable
+                of messages to be sent upon client connection.
+            handler (Callable[[bytes], Awaitable[AsyncIterable[bytes]]]): An
+                asynchronous message handler which returns an asynchronous iterable of
+                replies.
+        """
+
+        app = web.Application()
+
+        @routes.get("/")
+        async def handle(request):
+            return web.Response(text="Hello world!")
+
+        @routes.put("/command/foo/", name="command")
+        async def handle_put(request):  # self, data: int) -> str:
+
+            return web.Response(text=str("put data"))
+
+        @routes.get("/info/bar/{data}", name="info")
+        async def handle_get(request):  # self, json: Dict[str, Any]) -> None:
+            return web.Response(text="Your data: {}".format(request.match_info["data"]))
+
+        app.add_routes(routes)
+
+        runner = web.AppRunner(app)
+        await runner.setup()
+        site = web.TCPSite(runner, host=self.host, port=self.port)
+        await site.start()
+
+        await asyncio.Event().wait()
```

Comment:
Is there any way we can avoid blocking forever? Similarly to raising an error, ideally the closure of the web app should result in a cascade of closures until something decides to handle it properly or tickit exits.

Reply:
Something like?
```python
return web.run_app(...)
```

Reply:
Can't use ```python web.run_app()``` as that conflicts with Garry's implementation of asyncio

Reply:
```python
    try:
        asyncio.set_event_loop(loop)
        loop.run_until_complete(main_task)
    except (GracefulExit, KeyboardInterrupt):  # pragma: no cover
        pass
    finally:
        _cancel_tasks({main_task}, loop)
        _cancel_tasks(asyncio.all_tasks(loop), loop)
        loop.run_until_complete(loop.shutdown_asyncgens())
        loop.close()
        asyncio.set_event_loop(None)
```
This is in the `run_app()` function, i could just copy and paste this? :sweat_smile:

Reply:
Think we're discussing different things here, how we call `site.start()` is perfectly fine in my eyes. I merely want the `run_forever` method to block until and only until the `aiohttp` site is no longer online. I.e. no infinite blocking with `await asyncio.Event().wait()`

Reply:
I think using `run_app()` will do that, but still not sure why we can't. Can discuss tomorrow.

Reply:
Apparently ```python
web.run_app()
```
calls `loop.run_until_complete()` anyway so we can't call it here